### DISCUSSION TITLE
[Feature discussion] Translations & i18n

### DISCUSSION THREAD

**Discussion by rafaelurben** [rafaelurben/django-kmuhelper#197] [General]
Note: Translation progress is tracked in #198

**Comment by rafaelurben** [comment]
@crinulix I have seen that you recently forked this repo to https://github.com/crinulix/django-kmuhelper with the description "An accounting webapp with QR-bills for Swiss SMEs - French Version WIP"

Would you be interested in working together on this translation feature?

**Comment by crinulix** [comment]
Hi Rafael

Thank you for reaching out that's really kind of you. I must say that I'm
impressed by the amount of work you produced with KMU helper.

Yes, i forked your repo after testing it. Since it is completely in
German... I struggled to understand the business logic of the application.

I found that your application is really a good match/source for the
business that why wife is starting. We live in Geneva and her business for
the moment is targeting only the "Suisse Romande" region. So translating
the application to French was a natural decision to take, because, the app
users will be french speakers.

I also wanted to tweak certain aspects of the application so that it
matches the business needs of my wife. Since this business is a medical
service they are  actually exempt from paying VAT (Mehrwertsteuer/MwSt) so
that portion wouldn't be needed ( same for WooCommerce and goods delivery).
Also I wanted to have an appointment feature so that the medical service is
linked to specific appointment date.

Just as a note, I'm not a German speaker, and not a true native French
speaker.

So, as an exercise for me to better understand KMU helper code I was going
to translate everything to French. The funny thing happened when i started
inspecting the code and noticed that the majority of the variables are
defined in German. Because I was never exposed German before in my life (
I'm not of Swiss origin) i realized that for me to better understand the
code I actually needed to translate all of it to English, before attempting
to translate it to French...

So at the moment I'm finalizing a local translation of all KMU HELPER
repository into English, so that I can turn add/remove the features
required for medical services and then I'll finally translate it
French...so an English version is coming before a french one ...

So this is where I stand at. I really like your QR bill implementation and
that's why I'm using KMU HELPER. And yes help would be great 😃 to
translate it to French, but I also wanted to adapt certain features for
medical services purposes.... What are your plans for KMU helper? Do you
have many companies using it at the moment?


All the best.

**Comment by rafaelurben** [comment]
Hi @crinulix 

Thank you very much for such a detailed answer!

The mix between English and German in this repo has been bugging me since the beginning... 😂 

### Background

Because I and the target business owner speak German, this language was most important/natural to me. Because of that, I had started defining stuff in German before I realized that it might not have been the best idea...

But because the target business's customers are not all German-speaking, the generated QR invoice can be translated into German, French, Italian and English. (Mostly because these were the languages which had defined strings for the QR invoice - see [style guide](https://www.six-group.com/dam/download/banking-services/standardization/qr-bill/style-guide-qr-bill-en.pdf) page 31)

I have since made sort of a plan in my head on how I wanted to proceed:

### The translation plan

1) Use gettext functions and template tags (see [here](https://docs.djangoproject.com/en/4.0/topics/i18n/translation/)) for every string displayed in the interface. -> I just very recently made some changes to prepare for this.

(For this step, it would have to be discussed if German should stay the base language for translations or not.)

With this first step, the interface would already be usable for non-German speakers.
But because the code would still consist of a big language mess, I also planned more, less important changes:

2) Translate non-database-related variables
3) Translate database fields (need to be very careful here not to mess up migrations)
4) Translate the model names (again, very careful - this also requires some changes to admin URLs)

And at last, even less important:

5) Update the documentation (which is outdated anyway)

### My vision

The KMUHelper is currently only being used by the one business I created it for - at least as far as I know. Because this is a project I have dedicated a LOT of my free time to, I would be very happy if we could arrange it so you can use this for your wife's business.

You talked about features not being needed. If I/we could integrate these into the settings page or config file, this would make the KMUHelper more adaptable and would make it useful for more businesses, which is definitely something I'd look forward to.

I think it would be a waste of resources if both our versions would be diverging into very specific applications. Instead, I'd prefer if we found a way to incorporate everything you need into this version and make it modular with settings.

### How about you?

How do you see the future of KMUHelper? I'm looking forward to hearing your opinion on this!

\- Rafael

PS: I'm sorry if there are mistakes in my English, I am not a native English speaker. If there should be anything you don't understand, feel free to ask.
PPS: Feel free to contact me via email at kmuhelper@rafaelurben.ch if you prefer to talk more privately.
